Disallow empty statements made by semicolons

Burgos opened the pull request:
This turns the previous warning when using a semicolon
for an empty statement to a compilation error.

Burgos committed:
Fix compilation errors caused by empty statement in tests

There were places in the tests where the empty statements
were produces, which is now a compilation error. This removes/replaces
all of them with {}.

Burgos commented on src/ddmd/parse.d:
```diff
@@ -4664,7 +4664,7 @@ final class Parser(AST) : Lexer
             if (f.frequire || f.fensure)
                 error("missing body { ... } after in or out");
-            f.fbody = parseStatement(PSsemi);
+            f.fbody = parseStatement(0);
```
I was going to make a new enum value `PSdefault` or so, but I saw `parseStatement` called with `0` in other places as well, so I left it like this.

BBasile commented on src/ddmd/parse.d:
```diff
@@ -5112,10 +5112,7 @@ final class Parser(AST) : Lexer
-                    warning(loc, "use '{ }' for an empty statement, not a ';'");
-                else
-                    error("use '{ }' for an empty statement, not a ';'");
+                error("use '{ }' for an empty statement, not a ';'");
```
You cant pass from warning to error. There must be a  deprecation period before.

WalterBright commented:
This must go through a period of time where it is deprecated first.

Burgos commented:
Ah, true. I got mislead by the compiler's source, which states this is deprecated (See the `PSsemi`'s ddoc). Will fix, thanks!

Burgos committed:
Deprecate empty statements made by semicolon

Creating empty statements using semicolons, for which
was previously issued a warning is now causing a deprecation
hint.

Burgos commented:
Changed to a deprecation.

WalterBright reviewed (approved):
Nice work!

Burgos commented:
I'm not sure what's up with Jenkins, but he also failed in the trivial #6891 PR

wilzbach commented:
> I'm not sure what's up with Jenkins, but he also failed in the trivial #6891 PR

Yes, the Jenkins server was down for a week and now we are playing goose hunting for the introduced regressions in downstream projects, see e.g. https://github.com/dlang/phobos/pull/5469
tl;dr: don't worry about it, the errors should be fixed soon.

nemanja-boric-sociomantic commented:
All clear, thanks!
